File: app/src/main/java/com/dewonderstruck/apps/ashx0/viewmodel/product/TouchCountViewModel.java
package com.dewonderstruck.apps.ashx0.viewmodel.product;

import com.dewonderstruck.apps.ashx0.repository.product.ProductRepository;
import com.dewonderstruck.apps.ashx0.utils.AbsentLiveData;
import com.dewonderstruck.apps.ashx0.viewmodel.common.PSViewModel;
import com.dewonderstruck.apps.ashx0.viewobject.common.Resource;

import javax.inject.Inject;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.Transformations;

public class TouchCountViewModel extends PSViewModel {
    private final LiveData<Resource<Boolean>> sendTouchCountPostData;
    private MutableLiveData<TouchCountViewModel.TmpDataHolder> sendTouchCountDataPostObj = new MutableLiveData<>();

    @Inject
    TouchCountViewModel(ProductRepository productRepository) {
        sendTouchCountPostData = Transformations.switchMap(sendTouchCountDataPostObj, obj -> {

            if (obj == null) {
                return AbsentLiveData.create();
            }
            return productRepository.uploadTouchCountPostToServer(obj.userId, obj.typeId, obj.typeName);
        });
    }

    public void setTouchCountPostDataObj(String userId, String typeId, String typeName) {

        TouchCountViewModel.TmpDataHolder tmpDataHolder = new TouchCountViewModel.TmpDataHolder();
        tmpDataHolder.userId = userId;
        tmpDataHolder.typeId = typeId;
        tmpDataHolder.typeName = typeName;

        sendTouchCountDataPostObj.setValue(tmpDataHolder);

    }

    public LiveData<Resource<Boolean>> getTouchCountPostData() {
        return sendTouchCountPostData;
    }

    static class TmpDataHolder {
        public String userId = "";
        String typeId = "";
        String typeName = "";
        String shopId = "";
    }
}

File: app/src/main/java/com/dewonderstruck/apps/ashx0/viewmodel/product/TouchCountViewModel.kt
package com.dewonderstruck.apps.ashx0.viewmodel.product

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.Transformations
import com.dewonderstruck.apps.ashx0.repository.product.ProductRepository
import com.dewonderstruck.apps.ashx0.utils.AbsentLiveData
import com.dewonderstruck.apps.ashx0.viewmodel.common.PSViewModel
import com.dewonderstruck.apps.ashx0.viewobject.common.Resource
import javax.inject.Inject

class TouchCountViewModel @Inject internal constructor(productRepository: ProductRepository) : PSViewModel() {
    val touchCountPostData: LiveData<Resource<Boolean>>
    private val sendTouchCountDataPostObj = MutableLiveData<TmpDataHolder>()
    fun setTouchCountPostDataObj(userId: String, typeId: String, typeName: String) {
        val tmpDataHolder = TmpDataHolder()
        tmpDataHolder.userId = userId
        tmpDataHolder.typeId = typeId
        tmpDataHolder.typeName = typeName
        sendTouchCountDataPostObj.value = tmpDataHolder
    }

    internal class TmpDataHolder {
        var userId = ""
        var typeId = ""
        var typeName = ""
        var shopId = ""
    }

    init {
        touchCountPostData = Transformations.switchMap(sendTouchCountDataPostObj) { obj: TmpDataHolder? ->
            if (obj == null) {
                return@switchMap AbsentLiveData.create<Resource<Boolean>>()
            }
            productRepository.uploadTouchCountPostToServer(obj.userId, obj.typeId, obj.typeName)
        }
    }
}